src/vulkan/vertex_buffer.cc

```diff
@@ -75,11 +75,23 @@ Result VertexBuffer::SendVertexData(CommandBuffer* command) {
   if (!is_vertex_data_pending_)
     return Result("Vulkan::Vertices data was already sent");
 
-  for (const auto& buf : data_) {
+  for (size_t i = 0; i < data_.size(); i++) {
+    const auto& buf = data_[i];
+
+    // Search the previous bindings to see if the same Amber buffer is
+    // already bound.
+    for (size_t j = 0; j < i; j++) {
+      if (buf == data_[j]) {
+        // Found one. Reuse existing transfer buffer.
+        transfer_buffers_.emplace_back(transfer_buffers_[j]);
```

Comment:
What's the benefit of reusing vs just creating a new one? If we're reusing, does the initial transfer not copy the offset'd data? Why do we need to add it multiple times?

Reply:
The whole is copied regardless of the offset which is only used in descriptor write. Even if we would use the offset during the memcopy Amber wouldn't know how much to copy since it only knows the size of the whole buffer. We need to add the transfer buffer reference multiple times because the vector elements are later mapped directly to binding slots. In the end the same VkBuffer handle is being used for the descriptor write.

To be clear all this is for the sake of avoiding unnecessary memory copies. The implementation did work if I just disabled that one check for binding the buffer multiple times.

Reply:
If we just disabled that one check for binding the buffer multiple times, then wouldn't we actually end up creating and binding multiple VkBuffers, instead of binding one VkBuffer multiple times?

It is definitely a bit of pain for what seems like a simple feature but: I think we should avoid using shared_ptr. Is it possible to change this a bit, so that transfer_buffers_ is NOT used in `BindToCommandBuffer` to figure out the binding slots. `transfer_buffers_` could be a map from Buffer to TransferBuffer? Then instead loop over `data_` in `BindToCommandBuffer`?

Reply:
That's right. Using multiple VkBuffers would work but it's non-optimal.

I'll try using that map approach.